pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -85,6 +85,7 @@ type ManagerImpl struct {
 	// podDevices contains pod to allocated device mapping.
 	podDevices podDevices
 	store      utilstore.Store
+	pluginOpts map[string]*pluginapi.DevicePluginOptions
```

Comment:
This is needed here for handling container restarts whose endpoints have got deleted and not there in cache anymore.

Reply:
No it's not, the [next check](https://github.com/vikaschoudhary16/kubernetes/blob/e8b360fea64d916eff57cf10e89263dcaee4e453/pkg/kubelet/cm/devicemanager/manager.go#L691-L710) you do is to check if the endpoint exists...

Reply:
https://github.com/vikaschoudhary16/kubernetes/blob/e8b360fea64d916eff57cf10e89263dcaee4e453/pkg/kubelet/cm/devicemanager/manager.go#L698

consider the case where endpoint did not ask for prestart call and this endpoint got deleted from cache but there is a running container which used this endpoint. Now imagine restart of this container. In current code, [it will return nil](https://github.com/vikaschoudhary16/kubernetes/blob/e8b360fea64d916eff57cf10e89263dcaee4e453/pkg/kubelet/cm/devicemanager/manager.go#L698) and restart will be a success. container restart will fail only for those containers which were using endpoints that had asked for a prestart but now got deleted. If I do it like you are saying, irrespective of whether plugin asked for prestart or not, if endpoint is not in cache, container restart will fail. This is not the intended behavior.

Make sense?

Reply:
Ok that makes sense.

However it looks like we are not handling the path where a device plugin gets removed by the cluster administrator.
In that case pluginOpts never gets cleaned up

Reply:
Thats of no harm.